tests/0094-idempotence_msg_timeout.c

```diff
@@ -213,6 +215,84 @@ static void do_test_produce_timeout(const char *topic, const int msgrate) {
                  msgrate);
 }
 
+
+static void do_test_epoch_exhaustion(const char *topic, const int msgrate) {
+        rd_kafka_t *rk;
+        rd_kafka_conf_t *conf;
+        rd_kafka_topic_t *rkt;
+        rd_kafka_resp_err_t err;
+        sockem_ctrl_t ctrl;
+        int64_t initial_pid;
+        int64_t initial_epoch;
+        uint64_t testid;
+        const int partition = RD_KAFKA_PARTITION_UA;
+        int msgcnt          = msgrate * 10;
+        const int msgsize   = 100 * 1000;
+        int msgcounter      = 0;
+
+        testid = test_id_generate();
+
+        TEST_SAY(_C_BLU
+                 "Test idempotent producer epoch exhaustion");
+
+
+        test_conf_init(&conf, NULL, 60);
+        sockem_ctrl_init(&ctrl);
+
+        test_conf_set(conf, "enable.idempotence", "true");
+        test_conf_set(conf, "batch.size", "1");
+        test_conf_set(conf, "linger.ms", "1");
+
+        rd_kafka_conf_set_dr_msg_cb(conf, my_dr_msg_cb);
+
+        test_socket_enable(conf);
+        test_curr->is_fatal_cb = is_fatal_cb;
+
+        rk  = test_create_handle(RD_KAFKA_PRODUCER, conf);
+        rkt = test_create_producer_topic(rk, topic, "message.timeout.ms",
+                                         "100", NULL);
+
+        /* Create the topic to make sure connections are up and ready. */
+        err = test_auto_create_topic_rkt(rk, rkt, tmout_multip(5000));
+        TEST_ASSERT(!err, "topic creation failed: %s", rd_kafka_err2str(err));
+
+        initial_pid = rk->rk_eos.pid.id;
```

Comment:
Make sense, thank you. This is the guidance that I needed for the proper test!

Reply:
I am not sure if I understand the use of `rd_kafka_test_idemp_set_pid(..., int64_t id, int16_t epoch);` correctly. Am I able to set a valid PID from a client without mocking the broker?

Would it be okay to use `rd_kafka_test_idemp_set_pid_epoch(rd_kafka_t *rk, int16_t epoch);` that would se the epoch for the currently assigned PID with proper locking, and then retrieving the pid with `rd_kafka_test_get_pid(rk)`

>It would also be good if we could rely on the broker to verify that the epoch is correct and not wrapped, rather than just reading it back from rd_kafka_t which doesn't really prove correct system behaviour.

I will add a produce request that should not fail after this.